### PR TITLE
docs - update sql ref pages for ALTER/CREATE/DROP INDEX, REINDEX; rei…

### PR DESCRIPTION
**Pull request by lisakowen**
…ndex

this PR updates the listed sql command reference pages and the reindex utility to align with the postgres v12 ref page content.  there was some greenplum-specific content on the v6 pages, see notes/questions for more info.

opening this as a draft due to the number of outstanding questions.

notes/questions:  
- i don't know enough about the current index work-in-progress to know if/how that work may affect these reference pages.
- deferring partitioned table discussion to CREATE TABLE reference page as postgres docs.  note that that page has not yet been completed updated.
- looked to me like CONCURRENTLY is still not supported, so i didn't include any related info in the docs.  let me know if this was incorrect.
- i retained a GPORCA-specific statement in the description of method; look for XXX.  is this statement still true in greenplum 7?
- does hash=bitmap method?
- the page references the vacuum_cleanup_index_scale_factor, gin_pending_list_limit gucs.  we do not currently document these gucs - should we?
- does greenplum 7 support parallel index build for btree?  code says "GPDB_12_MERGE_FIXME: Parallel CREATE INDEX temporarily disabled."  i removed references to parallel index build from the docs.
- the CREATE page Notes section includes a statement about bitmap indexes - search for XXX.  is this statement still valid in greenplum 7?
- the CREATE page Notes section includes a statement about hash indexes - search for XXX.  is this statement still valid in greenplum 7?
- REINDEX PAGE notes include a few greenplum-specific statements - search for XXX.  are these still valid in greenplum 7?

doc review site links:
- ALTER INDEX (behind vpn), can navigate to others from there:  https://docs-staging.vmware.com/en/draft/VMware-Tanzu-Greenplum/sqlref-index2/greenplum-database/GUID-ref_guide-sql_commands-ALTER_INDEX.html
- reindexdb:  https://docs-staging.vmware.com/en/draft/VMware-Tanzu-Greenplum/sqlref-index2/greenplum-database/GUID-utility_guide-ref-reindexdb.html

### CODE REVIEW

File: gpdb-doc/markdown/ref_guide/sql_commands/CREATE_INDEX.html.md [hunk 1]
```diff
@@ -71,63 +88,87 @@ NULLS LAST
...
+For most index methods, the speed of creating an index is dependent on the setting of [maintenance\_work\_mem](../config_params/guc-list.html#maintenance_work_mem). Larger values will reduce the time needed for index creation, so long as you don't make it larger than the amount of memory really available, which would drive the machine into swapping.
 
-Columns with fewer than 100 distinct values usually do not benefit much from any type of index. For example, a gender column with only two distinct values for male and female would not be a good candidate for an index.
+XXX `bitmap` indexes perform best for columns that have between 100 and 100,000 distinct values. For a column with more than 100,000 distinct values, the performance and space efficiency of a bitmap index decline. The size of a bitmap index is proportional to the number of rows in the table times the number of distinct values in the indexed column.
```
Comment (huansong) [inline gpdb-doc/markdown/ref_guide/sql_commands/CREATE_INDEX.html.md:150]: This is correct in general but we might need to do more validation/tuning about the numbers. This is OK to keep for now.

**Comment by huansong** [comment]
For the questions:

> i don't know enough about the current index work-in-progress to know if/how that work may affect these reference pages.

Not that I'm aware of but I could be missing something. If there is, we should have follow-up doc stories to edit accordingly.

> looked to me like CONCURRENTLY is still not supported, so i didn't include any related info in the docs. let me know if this was incorrect.

Yes this is correct.

> does hash=bitmap method?

No they are two different index types.

> the page references the vacuum_cleanup_index_scale_factor, gin_pending_list_limit gucs. we do not currently document these gucs - should we?

Yes we should.

> does greenplum 7 support parallel index build for btree? code says "GPDB_12_MERGE_FIXME: Parallel CREATE INDEX temporarily disabled." i removed references to parallel index build from the docs.

Yes you are right - not supported.

>  i retained a GPORCA-specific statement in the description of method; look for XXX. is this statement still true in greenplum 7?
the CREATE page Notes section includes a statement about bitmap indexes - search for XXX. is this statement still valid in greenplum 7?
the CREATE page Notes section includes a statement about hash indexes - search for XXX. is this statement still valid in greenplum 7?
REINDEX PAGE notes include a few greenplum-specific statements - search for XXX. are these still valid in greenplum 7?

Answered in the code.

File: gpdb-doc/markdown/ref_guide/sql_commands/CREATE_INDEX.html.md [hunk 2]
```diff
@@ -71,63 +88,87 @@ NULLS LAST
...
-Prior releases of Greenplum Database also had an R-tree index method. This method has been removed because it had no significant advantages over the GiST method. If `USING rtree` is specified, `CREATE INDEX` will interpret it as `USING gist`.
+Use [DROP INDEX](DROP_INDEX.html) to remove an index.
 
-For more information on the GiST index type, refer to the [PostgreSQL documentation](https://www.postgresql.org/docs/9.4/indexes-types.html).
+Like any long-running transaction, `CREATE INDEX` on a table can affect which tuples can be removed by concurrent `VACUUM` on any other table.
 
-The use of hash indexes has been deactivated in Greenplum Database.
+Prior releases of Greenplum Database also had an R-tree index method. This method has been removed because it had no significant advantages over the GiST method. If `USING rtree` is specified, `CREATE INDEX` will interpret it as `USING gist`, to simplify conversion of old databases to GiST.
+
+XXX The use of hash indexes has been deactivated in Greenplum Database.
```
[the remark this answers is not in the document]
Comment (lisakowen) [reply gpdb-doc/markdown/ref_guide/sql_commands/CREATE_INDEX.html.md]: thank you for reviewing, @huansong .  i went ahead and removed this statement and updated the ddl topic (just to identify the list of types supported, i didn't update the complete topic.). will reinstate if it turns out we dont support it or there are some limitations.

File: gpdb-doc/markdown/ref_guide/sql_commands/REINDEX.html.md [hunk 3]
```diff
@@ -32,17 +36,24 @@ SYSTEM
...
 
-`REINDEX` does not update the `reltuples` and `relpages` statistics for the index. To update those statistics, run `ANALYZE` on the table after reindexing.
+XXX `REINDEX` does not update the `reltuples` and `relpages` statistics for the index. To update those statistics, run `ANALYZE` on the table after reindexing.
 
-If you suspect that shared global system catalog indexes are corrupted, they can only be reindexed in Greenplum utility mode. The typical symptom of a corrupt shared index is "index is not a btree" errors, or else the server crashes immediately at startup due to reliance on the corrupted indexes. Contact Greenplum Customer Support for assistance in this situation.
+XXX If you suspect that shared global system catalog indexes are corrupted, they can only be reindexed in Greenplum utility mode. The typical symptom of a corrupt shared index is "index is not a btree" errors, or else the server crashes immediately at startup due to reliance on the corrupted indexes. Contact Greenplum Customer Support for assistance in this situation.
```
[the remark this answers is not in the document]
Comment (lisakowen) [reply gpdb-doc/markdown/ref_guide/sql_commands/REINDEX.html.md]: PR to remove the paragraph from 6X_STABLE docs:  https://github.com/greenplum-db/gpdb/pull/14881

File: gpdb-doc/markdown/ref_guide/sql_commands/CREATE_INDEX.html.md [hunk 4]
```diff
@@ -71,65 +88,85 @@ NULLS LAST
...
+predicate
+:   The constraint expression for a partial index.
 
-:   `FASTUPDATE` - The GIN index method accepts the `FASTUPDATE` storage parameter. `FASTUPDATE` is a Boolean parameter that deactivates or enables the GIN index fast update technique. A value of ON enables fast update \(the default\), and OFF deactivates it. See [GIN fast update technique](https://www.postgresql.org/docs/12/gin-implementation.html#GIN-FAST-UPDATE) in the PostgreSQL documentation for more information.
+## <a id="section4isp"></a>Index Storage Parameters
 
-    > **Note** Turning `FASTUPDATE` off via `ALTER INDEX` prevents future insertions from going into the list of pending index entries, but does not in itself flush previous entries. You might want to VACUUM the table afterward to ensure the pending list is emptied.
+The optional `WITH` clause specifies *storage parameters* for the index. Each index method has its own set of allowed storage parameters. The B-tree, hash, GiST and SP-GiST index methods all accept this parameter:
```
Comment (huansong) [inline gpdb-doc/markdown/ref_guide/sql_commands/CREATE_INDEX.html.md:89]: `bitmap` supports `fillfactor` too

File: gpdb-doc/markdown/ref_guide/sql_commands/CREATE_INDEX.html.md [hunk 5]
```diff
@@ -71,65 +76,85 @@ NULLS LAST
...
 CREATE UNIQUE INDEX title_idx ON films (title);
 ```
 
-To create a bitmap index on the column `gender` in the table `employee`:
+To create a unique B-tree index on the column `title` with included columns `director` and `rating` in the table `films`:
 
 ```
-CREATE INDEX gender_bmp_idx ON employee USING bitmap 
-(gender);
+CREATE UNIQUE INDEX title_idx ON films (title) INCLUDE (director, rating);
```
Comment (huansong) [inline gpdb-doc/markdown/ref_guide/sql_commands/CREATE_INDEX.html.md:157]: need to remove/adjust this example as we don't have `INCLUDE` for now

File: gpdb-doc/markdown/ref_guide/sql_commands/REINDEX.html.md [hunk 6]
```diff
@@ -32,17 +36,22 @@ SYSTEM
...
 
-Reindexing a single index or table requires being the owner of that index or table. Reindexing a database requires being the owner of the database \(note that the owner can therefore rebuild indexes of tables owned by other users\). Of course, superusers can always reindex anything.
+Reindexing a single index or table requires being the owner of that index or table. Reindexing a schema or database requires being the owner of the schema or database. Note that it is therefore sometimes possible for non-superusers to rebuild indexes of tables owned by other users. However, as a special exception, when a non-superuser issues `REINDEX DATABASE`, `REINDEX SCHEMA` or `REINDEX SYSTEM`, Greenplum Database skips indexes on shared catalogs unless the user owns the catalog \(which typically won't be the case\). Of course, superusers can always reindex anything.
 
-`REINDEX` does not update the `reltuples` and `relpages` statistics for the index. To update those statistics, run `ANALYZE` on the table after reindexing.
+Greenplum Datase does not support reindexing partitioned tables or partitioned indexes. You can separately reindex each individual partition instead.
```
Comment (huansong) [inline gpdb-doc/markdown/ref_guide/sql_commands/REINDEX.html.md:52]: Greenplum supports it.

File: gpdb-doc/markdown/utility_guide/ref/reindexdb.html.md [hunk 7]
```diff
@@ -21,56 +22,77 @@ reindexdb -V | --version
...
--W \| --password
-:   Force a password prompt.
+-W
+--password
+:   Force `reindexdb` to prompt for a password before connecting to a database.
+:   This option is never essential, since `reindexdb` automatically prompts for a password if the server demands password authentication. However, `reindexdb` will waste a connection attempt finding out that the server wants a password. In some cases it is worth typing `-W` to avoid the extra connection attempt.
 
 --maintenance-db=dbname
-:   Specifies the name of the database to connect to discover what other databases should be reindexed. If not specified, the `postgres` database will be used, and if that does not exist, `template1` will be used.
+:   Specifies the name of the database to connect to discover what other databases should be reindexed, when `-a` or `--all` is used.. If not specified, the `postgres` database will be used, and if that does not exist, `template1` will be used. This can be a connection string. If so, connection string parameters will override any conflicting command line options. Also, connection string parameters other than the database name itself will be re-used when connecting to other databases.
```
Comment (huansong) [inline gpdb-doc/markdown/utility_guide/ref/reindexdb.html.md:95]: nit: `is used..`: extra `.`

**Review by huansong — APPROVED** [review]
Should be good after addressing the latest comments. Thanks!

File: gpdb-doc/markdown/ref_guide/sql_commands/REINDEX.html.md (hunk 6 above)
Comment (lisakowen) [reply gpdb-doc/markdown/ref_guide/sql_commands/REINDEX.html.md]: removed this statement.

**Comment by lisakowen** [comment]
thank you for answering all of the questions and the thorough review, @huansong .

**Pull request merged**